src/lustre/lustreAstNormalizer.ml

```diff
@@ -0,0 +1,346 @@
+  | Const (_, _) -> true
+  | _ -> false
+
+let normalize_list f list =
+  let over_list (nitems, gids) item =
+    let (normal_item, ids) = f item in
+    normal_item :: nitems, union ids gids
+  in List.fold_left over_list ([], empty) list
+
+let rec normalize (decls:LustreAst.t) =
+  let ast, gids = normalize_list normalize_declaration decls in
+  Log.log L_trace ("===============================================\n"
+    ^^ "Generated Identifiers:\n%a\n\n"
+    ^^ "Normalized lustre AST:\n%a\n"
+    ^^ "===============================================\n")
+    pp_print_generated_identifiers gids
+    A.pp_print_program ast;
+  Res.ok (ast, gids)
+
+and normalize_declaration = function
+  | NodeDecl (pos, decl) ->
+    let normal_decl, gids = normalize_node decl in
+    A.NodeDecl(pos, normal_decl), gids
+  | FuncDecl (pos, decl) ->
+    let normal_decl, gids = normalize_node decl in
+    A.FuncDecl (pos, normal_decl), gids
+  | decl -> decl, empty
+
+and normalize_node
+    (id, is_func, params, inputs, outputs, locals, items, contracts) =
+  let nitems, gids1 = normalize_list normalize_item items in
+  let ncontracts, gids2 = match contracts with
+    | Some contracts ->
+      let ncontracts, gids = normalize_list normalize_contract contracts in
+      (Some ncontracts), gids
+    | None -> None, empty
+  in let gids = union gids1 gids2 in
+  (id, is_func, params, inputs, outputs, locals, nitems, ncontracts), gids
+
+and normalize_item = function
+  | Body equation ->
+    let nequation, gids = normalize_equation equation in
+    Body nequation, gids
+  | AnnotMain b -> AnnotMain b, empty
+  | AnnotProperty (pos, name, expr) ->
+    let nexpr, gids = normalize_expr expr in
+    AnnotProperty (pos, name, nexpr), gids
+
+and normalize_contract = function
+  | Assume (pos, name, soft, expr) ->
+    let nexpr, gids = normalize_expr expr in
+    Assume (pos, name, soft, nexpr), gids
+  | Guarantee (pos, name, soft, expr) -> 
+    let nexpr, gids = normalize_expr expr in
+    Guarantee (pos, name, soft, nexpr), gids
+  | Mode (pos, name, requires, ensures) ->
+    let over_property (pos, name, expr) =
+      let nexpr, gids = normalize_expr expr in
+      (pos, name, nexpr), gids
+    in
+    let nrequires, gids1 = normalize_list over_property requires in
+    let nensures, gids2 = normalize_list over_property ensures in
+    Mode (pos, name, nrequires, nensures), union gids1 gids2
+  | ContractCall (pos, name, inputs, outputs) ->
+    let ninputs, gids = normalize_list normalize_expr inputs in
+    ContractCall (pos, name, ninputs, outputs), gids
+  | decl -> decl, empty
+
+and normalize_equation = function
+  | Assert (pos, expr) ->
+    let nexpr, gids = normalize_expr expr in
+    Assert (pos, nexpr), gids
+  | Equation (pos, lhs, expr) ->
+    let nexpr, gids = normalize_expr expr in
+    Equation (pos, lhs, nexpr), gids
+  | Automaton (pos, id, states, auto) ->
+    A.Automaton (pos, id, states, auto), empty
```

Comment:
Automaton requires normalization, it contains Lustre expressions too. Use a `Lib.todo __LOC__` call if you do not want to implement this now for some reason.